File: python/215.py
# Kth Largest Element in an Array
# 数组中的第K个最大元素


class Solution:
    def findKthLargest(self, nums: list[int], k: int) -> int:
        import heapq
        min_heap = []
        for num in nums:
            if len(min_heap) < k:
                heapq.heappush(min_heap, num)
            elif num > min_heap[0]:
                heapq.heappop(min_heap)
                heapq.heappush(min_heap, num)
        return min_heap[0]
    

if __name__ == '__main__':
    test_cases = [
        ([3,2,1,5,6,4], 2),
        ([3,2,3,1,2,4,5,5,6], 4),
    ]

    sol = Solution()
    for case in test_cases:
        res = sol.findKthLargest(*case)
        print(case, '\n', res)

File: cpp/215.cpp
// Kth Largest Element in an Array
// 数组中的第K个最大元素

#include "common.h"

class Solution
{
public:
    int findKthLargest(std::vector<int> &nums, int k)
    {
        std::priority_queue<int> max_heap;
        for (int i = 0; i < nums.size(); ++i)
        {
            if (i < k)
            {
                max_heap.emplace(-nums[i]);
            }
            else if (-nums[i] < max_heap.top())
            {
                max_heap.pop();
                max_heap.emplace(-nums[i]);
            }
        }
        return -max_heap.top();
    }
};

int main()
{
    std::vector<std::tuple<std::vector<int>, int>> test_cases = {
        {{3, 2, 1, 5, 6, 4}, 2},
        {{3, 2, 3, 1, 2, 4, 5, 5, 6}, 4},
    };

    Solution sol;
    for (auto &[nums, k] : test_cases)
    {
        int res = sol.findKthLargest(nums, k);
        std::cout << res << std::endl;
    }
}
